Add sample for OAuth Token via firebase functions

santhoshvaddi opened the pull request.

fhinkel commented:
👋 @santhoshvaddi Thanks for the PR. This repository is only for samples that we use via region tags in our documentation. This samples doesn't seem to have any tags. It might be better suited as part of a [community tutorial](https://cloud.google.com/community/).

nnegrey commented:
Hey, 

We're working with a Tech Writer to get the region tags in at the moment. 

Thanks.

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+admin.initializeApp(functions.config().firebase);
+var db = admin.firestore();
+ 
+ 
```
`// [START save_token_to_firebase]`

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+  var docRef = db.collection('DialogflowTokens').doc("OauthToken");
+  docRef.set(oauthToken);
+}
+ 
```
```
// [END save_token_to_firebase]

```

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+  return currentDate <= (expirationDate - 1000 * 60 * 5);
+}
+ 
+ 
```
`// [START function_get_token]`

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+      console.log('Error retrieving token', err);
+      return "Error retrieving token";
+    });
+});
```
`// [END function_get_token]`

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+}
+ 
+ 
+// This method verifies the token expiry by validating against current time
```
`// [START validate_token]`

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+  // If within 5 minutes of expiration, return false
+  return currentDate <= (expirationDate - 1000 * 60 * 5);
+}
+ 
```
`// [END validate_token]`

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+  });
+}
+ 
+ 
```
`// [START retrieve_credentials]`

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+    get_req.end();
+  });
+}
+ 
```
`// [END retrieve_credentials]`

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+  docRef.set(oauthToken);
+}
+ 
+ 
```
`// [START generate_token]`

nnegrey commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,149 @@
+    post_req.end();
+  });
+}
+ 
```
`// [END generate_token]`

ricalo commented:
Thanks for the suggested tags, @nnegrey and @santhoshvaddi. I'll provide some feedback on the exact location of the region comments when we have a good draft for the docs. Thanks for the guidance, @fhinkel.

santhoshvaddi commented:
Sure @nnegrey and @ricalo thanks for the update on this.

fhinkel commented:
Thanks you all! We need tests for these samples so that we can land them. You can follow the format of other tests in this repository. Please use the test framework `mocha`, we're in the process of refactoring all tests and there are still a few folders left to work on.

santhoshvaddi commented:
Thanks @fhinkel as suggested we have added tests using the mocha framework followed by other tests in this repository. Please let us know if need any changes required to it. @ricalo for now we have added the above @nnegrey suggested region tags, will change them once we have a good draft.

fhinkel commented on functions/dialogflow/functions/.eslintrc.json:
```diff
@@ -0,0 +1,123 @@
+{
```
You shouldn't need this file, we're using the top level config.

santhoshvaddi replied:
Removed.

fhinkel commented on functions/dialogflow/functions/.gitignore:
```diff
@@ -0,0 +1 @@
+node_modules/
```
Not needed, `node_modules` is excluded in the top level gitignore.

santhoshvaddi replied:
Removed.

fhinkel commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,154 @@
+/* eslint-disable promise/no-nesting */
```
This line should not be needed.

santhoshvaddi replied:
Removed.

fhinkel commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,154 @@
+ 
+// [START save_token_to_firebase]
+function saveOAuthToken(context, oauthToken) {
+  var docRef = db.collection('DialogflowTokens').doc("OauthToken");
```
`const` instead of `var`. Since you're using ES6 syntax, I'd always use `let/const` instead of `var`.

santhoshvaddi replied:
The functionality of var and let is different from in swift and javascript :) . Updated the code with let/const instead of var.

fhinkel commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,154 @@
+  // a new token for a 2nd service account that only has the permission to 
+  // act as a Dialogflow Client
+  return new Promise((resolve, reject) => {
+    var post_options = {
```
const

santhoshvaddi replied:
Changed to const

fhinkel commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,154 @@
+    };
+ 
+    // Set up the request
+    var oauthToken = '';
```
let

santhoshvaddi replied:
changed to let

fhinkel commented on functions/dialogflow/firebase.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "functions": {
+    "predeploy": [
```
Is this step necessary for this sample? Seems like a lot of overhead.

santhoshvaddi replied:
I will remove these lines.

fhinkel commented:
Is `functions/dialogflow/functions` the right path? @ace-n, do we keep Firebase functions samples in `functions`? I think it should either be `functions/dialogflow` or maybe `firebase-oauth`.

fhinkel commented:
Please add the config files for CI in `.kokoro`.

fhinkel commented:
@santhoshvaddi Please have a look at the linter issues. Thanks!

fhinkel commented:
Waiting for .kokoro config on presubmit so we can run the new tests before merging. Will restart tests once config landed.

santhoshvaddi commented on functions/dialogflow/functions/index.js:
```diff
@@ -0,0 +1,153 @@
+    });
+ 
+    // Sets up the scope that we want the end user to have.
+    const body = `{ 
```
Done.

fhinkel commented:
The functions/dialogflow tests are not passing. Please have a look. Thanks

nnegrey commented:
Hey Franzie, 

Quick question: will this testing framework work for Firebase Functions? 

Context: The sample here is part of a larger tutorial with an attached iOS app: https://github.com/GoogleCloudPlatform/ios-docs-samples/pull/105 that demonstrates how to call GCP APIs from mobile

fhinkel commented:
The tests should call the methods in the sample so we notice when they code is out of date. This works with `mocha` independently of the larger scope of the app.

fhinkel commented on functions/dialogflow/functions/package.json:
```diff
@@ -0,0 +1,28 @@
+    "test": "npm run local-test"
+  },
+  "dependencies": {
+    "firebase-admin": "~7.0.0",
```
Can we use `^7.0.0`?

santhoshvaddi replied:
I am not exactly sure on this @fhinkel but may be we can.

fhinkel commented:
Should I go ahead and merge it? 😄
